MapUtil.handle does not support Array objects

**I am using Json-B that implemented by Eclipse Yasson. I want to serialize my custom Map that there is a value is an instance of array of String. I got this error:**

Caused by: java.lang.IllegalArgumentException: Type class [Ljava.lang.String; is not supported.
        at org.eclipse.parsson.MapUtil.handle(MapUtil.java:75)
        at org.eclipse.parsson.JsonObjectBuilderImpl.populate(JsonObjectBuilderImpl.java:200)
        at org.eclipse.parsson.JsonObjectBuilderImpl.<init>(JsonObjectBuilderImpl.java:66)
        at org.eclipse.parsson.JsonProviderImpl.createObjectBuilder(JsonProviderImpl.java:187)
        at jakarta.json.Json.createObjectBuilder(Json.java:303)
        at com.appslandia.common.jose.JsonbMapAdapter.adaptToJson(JsonbMapAdapter.java:56)
        at com.appslandia.common.jose.JsonbMapAdapter.adaptToJson(JsonbMapAdapter.java:39)
        at org.eclipse.yasson.internal.serializer.AdaptedObjectSerializer.serialize(AdaptedObjectSerializer.java:62)

**I check on **parsson.MapUtil.handle** and I found out this method does NOT support array objects?**

**My suggession fix:**

    static JsonValue handle(Object value, JsonContext jsonContext) {
        if (value == null) {
            return JsonValue.NULL;
        
        // ... 
        } else if (value instanceof Array) {
            Collection<?> collection = toCollection(value);
            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(collection, jsonContext);
            return jsonArrayBuilder.build();
        }

        throw new IllegalArgumentException(String.format("Type %s is not supported.", value.getClass()));
    }

        static Collection<?> toCollection(Object arr) {
        	int len = Array.getLength(arr);
        	List<Object> list = new ArrayList<>(len);
        
        	for (int i = 0; i < len; i++) {
        	    list.add(Array.get(arr, i));
        	}
        	return list;
        }

If I use list of String instead array of String, my custom map was serialized fine.

Please notes that, If I use Gson or Jackson, my custom map was serialized fine.

--- a/impl/src/main/java/org/eclipse/parsson/MapUtil.java
+++ b/impl/src/main/java/org/eclipse/parsson/MapUtil.java
@@ -21,6 +21,7 @@ import jakarta.json.JsonObjectBuilder;
 import jakarta.json.JsonValue;
 import java.math.BigDecimal;
 import java.math.BigInteger;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Map;
 
@@ -67,6 +68,57 @@ public final class MapUtil {
             @SuppressWarnings("unchecked")
             JsonObjectBuilder object = new JsonObjectBuilderImpl((Map<String, Object>) value, jsonContext);
             return object.build();
+        } else if (value instanceof Object[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(Arrays.asList((Object[]) value), jsonContext);
+            return jsonArrayBuilder.build();
+        } else if (value instanceof int[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (int d : (int[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof long[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (long d : (long[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof double[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (double d : (double[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof boolean[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (boolean d : (boolean[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof char[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (char d : (char[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof float[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (float d : (float[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof byte[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (byte d : (byte[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
+        } else if (value instanceof short[]) {
+            JsonArrayBuilder jsonArrayBuilder = new JsonArrayBuilderImpl(jsonContext);
+            for (short d : (short[]) value) {
+                jsonArrayBuilder.add(d);
+            }
+            return jsonArrayBuilder.build();
         }
 
         throw new IllegalArgumentException(String.format("Type %s is not supported.", value.getClass()));

--- a/impl/src/test/java/org/eclipse/parsson/tests/JsonObjectTest.java
+++ b/impl/src/test/java/org/eclipse/parsson/tests/JsonObjectTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2012, 2021 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012, 2023 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -229,4 +229,23 @@ public class JsonObjectTest extends TestCase {
         assertTrue(object3.hashCode() == object4.hashCode()); //equal instances have same hashCode
     }
 
+    public void testArrays() {
+        String[] stringArr = new String[] {"a", "b", "c"};
+        boolean[] boolArr = new boolean[] {true, false, true};
+        int[] intArr = new int[] {1, 2, 3};
+        char[] charArr = new char[] {'a', 'b', 'c'};
+        float[] floatArr = new float[] {1.0f, 2.0f, 3.0f};
+        Map<String, Object> m = new HashMap<>();
+        m.put("stringArray", stringArr);
+        m.put("booleanArray", boolArr);
+        m.put("intArray", intArr);
+        m.put("charArray", charArr);
+        m.put("floatArray", floatArr);
+        JsonObject object = Json.createObjectBuilder(m).build();
+        assertEquals("b", object.get("stringArray").asJsonArray().getString(1));
+        assertEquals(false, object.get("booleanArray").asJsonArray().getBoolean(1));
+        assertEquals(2, object.get("intArray").asJsonArray().getInt(1));
+        assertEquals('b', object.get("charArray").asJsonArray().getInt(1));
+        assertEquals(2.0, object.get("floatArray").asJsonArray().getJsonNumber(1).doubleValue());
+    }
 }

